src/zarrman/fetcher.rs

```diff
@@ -0,0 +1,188 @@
+use super::consts::{MANIFEST_CACHE_IDLE_EXPIRY, MANIFEST_ROOT_URL};
+use super::manifest::Manifest;
+use super::resources::ManifestPath;
+use super::util::{Index, ZarrManError};
+use crate::httputil::{BuildClientError, Client, HttpError, HttpUrl};
+use crate::paths::PureDirPath;
+use get_size::GetSize;
+use moka::{
+    future::{Cache, CacheBuilder},
+    ops::compute::{CompResult, Op},
+};
+use serde::Serialize;
+use std::sync::Arc;
+use std::time::Duration;
+
+/// A client for fetching & caching data from the manifest tree
+#[derive(Clone, Debug)]
+pub(crate) struct ManifestFetcher {
+    /// The HTTP client used for making requests to the manifest tree
+    inner: Client,
+
+    /// A cache of parsed manifest files, keyed by their path under
+    /// `MANIFEST_ROOT_URL`
+    cache: Cache<ManifestPath, Arc<Manifest>>,
+
+    /// [`MANIFEST_ROOT_URL`], parsed into an [`HttpUrl`]
+    manifest_root_url: HttpUrl,
+}
+
+impl ManifestFetcher {
+    /// Construct a new client instance
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if construction of the inner `reqwest::Client` fails
+    pub(crate) fn new(cache_size: u64) -> Result<Self, BuildClientError> {
+        let inner = Client::new()?;
+        let cache: Cache<ManifestPath, Arc<Manifest>> = CacheBuilder::new(cache_size)
+            .name("zarr-manifests")
+            .weigher(|_, manifest: &Arc<Manifest>| {
+                u32::try_from(manifest.get_size()).unwrap_or(u32::MAX)
+            })
+            .time_to_idle(MANIFEST_CACHE_IDLE_EXPIRY)
+            .eviction_listener(|path, manifest, cause| {
+                tracing::debug!(
```

Comment:
test coverage annotation is funny here but suggests that most here is not tested Does it mean that logging on eviction doesn't happen ie. evictions do not happen during testing?  (I am not yet a good reader of Rust, sorry)

Reply:
None of the caching is currently tested.